CRM/Queue/Queue/Sql.php

```diff
@@ -47,31 +47,27 @@ public function claimItem($lease_time = NULL) {
     $lease_time = $lease_time ?: $this->getSpec('lease_time') ?: static::DEFAULT_LEASE_TIME;
 
     $result = NULL;
-    $dao = CRM_Core_DAO::executeQuery('LOCK TABLES civicrm_queue_item WRITE;');
-    $sql = "
+    CRM_Core_DAO::executeQuery('LOCK TABLES civicrm_queue_item WRITE;');
+    $sql = '
         SELECT first_in_queue.* FROM (
           SELECT id, queue_name, submit_time, release_time, run_count, data
           FROM civicrm_queue_item
           WHERE queue_name = %1
-          ORDER BY weight ASC, id ASC
+          ORDER BY weight, id
           LIMIT 1
         ) first_in_queue
         WHERE release_time IS NULL OR release_time < %2
-      ";
+      ';
     $params = [
       1 => [$this->getName(), 'String'],
       2 => [CRM_Utils_Time::getTime(), 'Timestamp'],
     ];
     $dao = CRM_Core_DAO::executeQuery($sql, $params, TRUE, 'CRM_Queue_DAO_QueueItem');
-    if (is_a($dao, 'DB_Error')) {
-      // FIXME - Adding code to allow tests to pass
-      throw new CRM_Core_Exception('Unable to claim queue item');
-    }
```

Comment:
> The slightly meaningful thing in there is not handling the possibility it might not return an exception - if we are not setting the error scope (which we are not doing by passing in the relevant variable) then it won't return an error object

As I understand, `executeQuery` abides by PEAR error mode, and it can emit errors per one these policies:

1. Throw an `Exception`
    * If this is the active mode, then the guard above is not called. Rather, it emits a similar exception (which should be good enough for our purposes).
    * This is generally the more desirable mode.
    * Per `CRM_Core_Config::singleton()`, it appears that this is the default-global mode (once Civi is running).
2. Return an error object. (*Old school PEAR default. The code above checks that case.*)
    * If this is the active mode, then the guard above does something useful.
    * IIRC, this is the official default if you don't boot Civi. (*But who would do that...?*)
3. Show Civi's fatal error screen
    * If this is the active mode, then the guard above is not called (because it fatals/exits).
4. Ignore the exception
    * If this is the active mode, then the guard above is not called (because it's ignored).

IMHO, the guard here is only about mode `#2`. If that doesn't happen anymore, then 🎉 and +1 for removing the guard.

Reply:
@totten 2 should only happen when we have set to `ignoreException` per the screenshot - we don't really do that any more - & not in any of the paths relevant here IMHO

Reply:
Agree none of the `ignoreException` cases look relevant to queue-claiming.

Equating `DB_Error` and `ignoreException` use-cases was a bit counter-intuitive, so I played with it, and it basically checks out. More specifically:

* If you set `CRM_Core_TemporaryErrorScope::ignoreException` (aka register `CRM_Core_Error::nullHandler`), then it does return `DB_Error`s.
* You can get `DB_Error`s without `ignoreException` (e.g. in absence of a global handler, it returns `DB_Error`s).
* But in practice, there is always a global handler, and there doesn't seem to be anything besides  `CRM_Core_TemporaryErrorScope::ignoreException`  that has the effect of switching the active policy (`_PEAR_default_error...`).